tests/turnstile_complex.c

```diff
@@ -0,0 +1,171 @@
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      sched_unlend_prio(td, prio);                                             \
+    }                                                                          \
+  } while (0)
+
+#define set_base_prio(td, prio)                                                \
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      td->td_base_prio = prio;                                                 \
+    }                                                                          \
+  } while (0)
+
+static mtx_t *mtx1 = &MTX_INITIALIZER(MTX_DEF);
+static mtx_t *mtx2 = &MTX_INITIALIZER(MTX_DEF);
+static thread_t *td[T];
+static volatile bool low_prio_mtx1_acquired;
+static volatile bool med_prio_mtx1_acquired;
+static volatile bool high_prio_mtx2_acquired;
+
+enum {
+  /* Priorities are multiplies of RunQueue_PriorityPerQueue
+   * so that each priority matches different run queue. */
+  LOW = 0,
+  MED = RQ_PPQ,
+  HIGH = 2 * RQ_PPQ
+};
+
+/* td3 */
+static void high_prio_task(void *arg) {
+  WITH_MTX_LOCK (mtx2) {
+    /* When we get here, low_prio_task and med_prio_task1 should've
+       got their mutexes and all priorities should be in initial state. */
+    high_prio_mtx2_acquired = 1;
+    assert(low_prio_mtx1_acquired == 1);
+    assert(med_prio_mtx1_acquired == 1);
+
+    assert_priorities(LOW, MED, MED, HIGH);
+  }
+}
+
+/* td1 */
+static void med_prio_task1(void *arg) {
+  WITH_NO_PREEMPTION {
+    WITH_MTX_LOCK (mtx2) {
+      WITH_MTX_LOCK (mtx1) {
+        /* Td0 (low_prio_task) has released mtx1.
+         * Td3 (high_prio_task) is waiting for mtx2. */
+        med_prio_mtx1_acquired = 1;
+        assert(low_prio_mtx1_acquired == 1);
+        assert(high_prio_mtx2_acquired == 0);
+
+        assert(thread_self()->td_prio == HIGH);
+        assert(thread_self()->td_flags & TDF_BORROWING);
+      }
+    }
+  }
+
+  /* Enabling preemption should've preempted us and switched to
+   * high_prio_task. */
+  assert(high_prio_mtx2_acquired == 1);
+  assert(!(thread_self()->td_flags & TDF_BORROWING));
+  assert_priorities(LOW, MED, MED, HIGH);
+}
+
+/* td2 */
+static void med_prio_task2(void *arg) {
+  /* Without turnstile mechanism this assert would fail. */
+  assert(high_prio_mtx2_acquired);
+}
+
+/* td0 */
+static void low_prio_task(void *arg) {
+  WITH_NO_PREEMPTION {
+    unlend_prio(td[0], LOW);
+    lend_prio(td[1], MED);
+
+    /* Thread td1 (med_prio_task1) will run first after next thread_yield(). */
+    assert_priorities(LOW, MED, LOW, LOW);
+
+    WITH_MTX_LOCK (mtx1) {
+      low_prio_mtx1_acquired = 1;
+      assert(med_prio_mtx1_acquired == 0);
+      assert(high_prio_mtx2_acquired == 0);
+
+      thread_yield();
```

Comment:
What about `assert(turnstile_lookup(mtx1) == NULL);`? It would tell us that `mtx1` was not contested when `td0` acquired it, so nothing could happen with priorities.

Reply:
I actually intended it as a reminder about priorities of the threads - so that the reader will immediately know which thread will run after `thread_yield();`. It is indeed written a few lines above but I think repeating it here would make it easier to read the code.